mmseg/models/decode_heads/protoseg_head.py

```diff
@@ -0,0 +1,156 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+# from mmseg.models.backbones import HRNet
+import os
+import mmcv
+
+import functools
+import os
+import pdb
+
+import torch
+import torch.nn as nn
+from torch.nn.functional import interpolate
+from einops import rearrange, repeat
+from mmseg.registry import MODELS
+
+
+class ModuleHelper(object):
+
+    @staticmethod
+    def BNReLU(num_features, bn_type=None, **kwargs):
+        if bn_type == 'torchbn':
+            return nn.Sequential(
+                nn.BatchNorm2d(num_features, **kwargs),
+                nn.ReLU()
+            )
+        elif bn_type == 'torchsyncbn':
+            return nn.Sequential(
+                nn.SyncBatchNorm(num_features, **kwargs),
+                nn.ReLU()
+            )
+        elif bn_type == 'gn':
+            return nn.Sequential(
+                nn.GroupNorm(num_groups=8, num_channels=num_features, **kwargs),
+                nn.ReLU()
+            )
+        elif bn_type == 'fn':
+            exit(1)
+
+        else:
+            exit(1)
+            
+def l2_normalize(x):
+    return F.normalize(x, p=2, dim=-1)
+
+class ProjectionHead(nn.Module):
+    def __init__(self, dim_in, proj_dim=256, proj='convmlp', bn_type='torchsyncbn'):
+        super(ProjectionHead, self).__init__()
+
+        # Log.info('proj_dim: {}'.format(proj_dim))
+
+        if proj == 'linear':
+            self.proj = nn.Conv2d(dim_in, proj_dim, kernel_size=1)
+        elif proj == 'convmlp':
+            self.proj = nn.Sequential(
+                nn.Conv2d(dim_in, dim_in, kernel_size=1),
+                ModuleHelper.BNReLU(dim_in, bn_type=bn_type),
+                nn.Conv2d(dim_in, proj_dim, kernel_size=1)
+            )
+
+    def forward(self, x):
+        return F.normalize(self.proj(x), p=2, dim=1)
+
+class Protoseg_head(nn.Module):
+    def __init__(self,in_channels):
+        super(Hrnet,self).__init__()
+        self.in_channel=in_channels
+        self.num_prototype=10
+        self.update_prototype=True
+        self.gamma=0.999
+        self.pretrain_prototype=False
+        self.num_classes=19
+        self.cls_head = nn.Sequential(
+            nn.Conv2d(in_channels, in_channels, kernel_size=3, stride=1, padding=1),
+            ModuleHelper.BNReLU(in_channels, bn_type="torchbn"),
+            nn.Dropout2d(0.10)
+        )
+        
+        self.prototypes = nn.Parameter(torch.zeros(self.num_classes, self.num_prototype, in_channels),
+                                       requires_grad=True)
+
+        self.proj_head = ProjectionHead(in_channels, in_channels)
```

Comment:
Since `ProjectionHead` is just a conv2d, the class `ProjectionHead` is not needed.